File: 2019/rust/src/passwords.rs
use std::iter::FromIterator;
use std::collections::HashMap;

fn increasing(password: &str) -> bool {
    let mut sorted: Vec<char> = password[..].chars().collect();
    sorted.sort();

    password == String::from_iter(sorted)
}

fn cardinality(s: &str) -> HashMap<char,u32> {
    let mut counter: HashMap<char, u32> = HashMap::new();
    for c in s.chars() {
        match counter.insert(c, 1) {
            None    => {},
            Some(n) => { counter.insert(c, n + 1); },
        }
    }

    counter
}

fn duplicates(password: &str) -> bool {
    let mut count: u32 = 0;
    for (_key, val) in cardinality(password).iter() {
        if val > &1 { count += 1; }
    }

    count > 0
}

fn twins(password: &str) -> bool {
    let mut count: u32 = 0;
    for (_key, val) in cardinality(password).iter() {
        if val == &2 { count += 1; }
    }

    count > 0

}

pub fn solve(data: String) {

    let start: u32 = data[0..6].parse::<u32>().expect("invalid input");
    let end: u32 = data[7..].parse::<u32>().expect("invalid input");

    let candidates: Vec<String> = (start..end)
        .map(|p| p.to_string())
        .collect();

    let valid_passwords = candidates
        .iter()
        .filter(|p| increasing(p) && duplicates(p))
        .collect::<Vec<&String>>();

    println!("Part 1: {}", valid_passwords.len());

    let valid_passwords = valid_passwords
        .iter()
        .filter(|p| twins(p))
        .collect::<Vec<&&String>>();
    println!("Part 2: {}", valid_passwords.len());
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn passwords_increasing() {
        assert_eq!(increasing("111111"), true);
        assert_eq!(increasing("223450"), false);
        assert_eq!(increasing("123789"), true);
    }

    #[test]
    fn passwords_duplicates() {
        assert_eq!(duplicates("111111"), true);
        assert_eq!(duplicates("223450"), true);
        assert_eq!(duplicates("123789"), false);
    }

    #[test]
    fn passwords_twins() {
        assert_eq!(twins("112233"), true);
        assert_eq!(twins("123444"), false);
        assert_eq!(twins("111122"), true);
    }
}

File: 2019/haskell/src/Passwords.hs
module Passwords (valid1, valid2, day04) where

import           Control.Monad   (liftM2)
import           Data.List       (sort)
import           AdventOfCode    (cardinality)
import qualified Data.Map.Strict as M

increasing :: String -> Bool
increasing password = password == (sort password)

duplicates :: String -> Bool
duplicates password = (length candidates) > 0
  where candidates = filter (> 1) (M.elems $ cardinality password)

twins :: String -> Bool
twins password = (length candidates) > 0
  where candidates = filter (== 2) (M.elems $ cardinality password)

valid1 :: String -> Bool
valid1 = liftM2 (&&) increasing duplicates

valid2 :: String -> Bool
valid2 = liftM2 (&&) increasing twins

day04 :: [Int] -> IO ()
day04 (start:end:_) = do
  let passwords = map show [start ..end - 1]
  print $ length $ filter valid1 passwords
  print $ length $ filter valid2 passwords
